File: app/src/main/java/com/rollapp/shared/ui/members/MembersScreen.kt
package com.rollapp.shared.ui.members

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.ArrowBack
import androidx.compose.material.icons.rounded.MoreVert
import androidx.compose.material.icons.rounded.PersonAdd
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.rollapp.shared.domain.model.Member
import com.rollapp.shared.ui.components.InlineError
import com.rollapp.shared.ui.components.Sharing
import com.rollapp.shared.ui.components.UserAvatar

@Composable
fun MembersScreen(
    onBack: () -> Unit,
    viewModel: MembersViewModel = hiltViewModel()
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    val context = LocalContext.current
    var pendingRemoval by remember { mutableStateOf<Member?>(null) }

    pendingRemoval?.let { member ->
        AlertDialog(
            onDismissRequest = { pendingRemoval = null },
            title = { Text("Remove ${member.name}?") },
            text = {
                Text(
                    "They'll lose access to this group's photos. Photos they already " +
                        "uploaded stay in the group."
                )
            },
            confirmButton = {
                TextButton(onClick = {
                    viewModel.removeMember(member.uid)
                    pendingRemoval = null
                }) { Text("Remove") }
            },
            dismissButton = {
                TextButton(onClick = { pendingRemoval = null }) { Text("Cancel") }
            }
        )
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Members") },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(Icons.Rounded.ArrowBack, contentDescription = "Back")
                    }
                }
            )
        }
    ) { padding ->
        LazyColumn(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding),
            contentPadding = PaddingValues(bottom = 32.dp)
        ) {
            state.error?.let { error ->
                item(key = "error") {
                    InlineError(
                        message = error.message ?: "Something went wrong",
                        actionLabel = "Dismiss",
                        onAction = viewModel::dismissError,
                        modifier = Modifier.padding(16.dp)
                    )
                }
            }

            state.group?.takeIf { it.isInviteActive }?.let { group ->
                item(key = "invite") {
                    OutlinedButton(
                        onClick = {
                            Sharing.shareInvite(context, group.name, group.inviteCode, group.inviteLink)
                        },
                        shape = MaterialTheme.shapes.medium,
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(horizontal = 16.dp, vertical = 12.dp)
                            .height(50.dp)
                    ) {
                        Icon(Icons.Rounded.PersonAdd, contentDescription = null, modifier = Modifier.size(18.dp))
                        Spacer(Modifier.width(8.dp))
                        Text("Invite friends · ${group.inviteCode}")
                    }
                }
            }

            items(state.members, key = { it.uid }) { member ->
                MemberRow(
                    member = member,
                    isMe = member.uid == state.myUid,
                    canManage = state.isAdmin && member.uid != state.myUid,
                    onRemove = { pendingRemoval = member }
                )
            }
        }
    }
}

@Composable
private fun MemberRow(
    member: Member,
    isMe: Boolean,
    canManage: Boolean,
    onRemove: () -> Unit
) {
    var menuOpen by remember { mutableStateOf(false) }

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 10.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        UserAvatar(
            name = member.name,
            photoUrl = member.photoUrl,
            seed = member.uid,
            size = 44.dp
        )
        Spacer(Modifier.width(14.dp))

        Column(modifier = Modifier.weight(1f)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    text = if (isMe) "${member.name} (you)" else member.name,
                    style = MaterialTheme.typography.titleMedium
                )
                if (member.isAdmin) {
                    Spacer(Modifier.width(8.dp))
                    Box(
                        modifier = Modifier
                            .clip(MaterialTheme.shapes.extraSmall)
                            .background(MaterialTheme.colorScheme.primaryContainer)
                            .padding(horizontal = 7.dp, vertical = 2.dp)
                    ) {
                        Text(
                            text = "Admin",
                            style = MaterialTheme.typography.labelSmall,
                            color = MaterialTheme.colorScheme.onPrimaryContainer
                        )
                    }
                }
            }
            Text(
                text = "${member.photoCount} ${if (member.photoCount == 1) "photo" else "photos"}",
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }

        if (canManage) {
            Box {
                IconButton(onClick = { menuOpen = true }) {
                    Icon(Icons.Rounded.MoreVert, contentDescription = "Manage ${member.name}")
                }
                DropdownMenu(expanded = menuOpen, onDismissRequest = { menuOpen = false }) {
                    DropdownMenuItem(
                        text = { Text("Remove from group") },
                        onClick = { menuOpen = false; onRemove() }
                    )
                }
            }
        }
    }
}

File: ios/WeWere/UI/Screens/MembersScreen.swift
import SwiftUI
import Combine

struct MembersUiState: Equatable {
    var group: Group? = nil
    var members: [Member] = []
    var myUid: String? = nil
    var isAdmin: Bool = false
    var error: AppError? = nil
    var isLoading: Bool = true
}

@MainActor
final class MembersViewModel: ObservableObject {
    @Published private(set) var state = MembersUiState()

    let groupId: String
    private let groupRepository: GroupRepository
    private let errorFlow = CurrentValueSubject<AppError?, Never>(nil)
    private var cancellables = Set<AnyCancellable>()

    init(groupId: String, container: AppContainer = .shared) {
        self.groupId = groupId
        groupRepository = container.groupRepository
        let authRepository = container.authRepository

        groupRepository.observeGroup(groupId: groupId)
            .combineLatest(groupRepository.observeMembers(groupId: groupId),
                           groupRepository.observeMembership(groupId: groupId), errorFlow)
            .map { group, members, membership, error in
                MembersUiState(
                    group: group,
                    // Most photos first — on a trip this is genuinely the interesting ranking.
                    members: members.sorted { $0.photoCount > $1.photoCount },
                    myUid: authRepository.currentUid(),
                    isAdmin: membership?.isAdmin == true,
                    error: error,
                    isLoading: false
                )
            }
            .receive(on: DispatchQueue.main)
            .sink { [weak self] in self?.state = $0 }
            .store(in: &cancellables)
    }

    func removeMember(_ uid: String) {
        Task {
            if case .failure(let error) = await groupRepository.removeMember(groupId: groupId, uid: uid) {
                errorFlow.send(error)
            }
        }
    }

    func regenerateInvite() {
        Task {
            if case .failure(let error) = await groupRepository.regenerateInviteCode(groupId: groupId, expiresAt: nil) {
                errorFlow.send(error)
            }
        }
    }

    func dismissError() { errorFlow.send(nil) }
}

struct MembersScreen: View {
    let onBack: () -> Void

    @StateObject private var viewModel: MembersViewModel
    @EnvironmentObject private var overlays: Overlays

    init(groupId: String, onBack: @escaping () -> Void) {
        self.onBack = onBack
        _viewModel = StateObject(wrappedValue: MembersViewModel(groupId: groupId))
    }

    var body: some View {
        let state = viewModel.state
        VStack(spacing: 0) {
            RollTopBar(title: "Members", onBack: onBack)
            ScrollView {
                LazyVStack(spacing: 0) {
                    if let error = state.error {
                        InlineError(message: error.message ?? "Something went wrong",
                                    actionLabel: "Dismiss", onAction: viewModel.dismissError)
                            .padding(16)
                    }

                    if let group = state.group, group.isInviteActive {
                        VStack(spacing: 0) {
                            // Someone across the table scans this instead of typing.
                            QrCode(content: group.inviteLink, size: 180)
                            Spacer().frame(height: 10)
                            Text(group.inviteCode).rollStyle(RollType.headlineSmall, color: Gold)
                            Text("Scan this, or type the code").rollStyle(RollType.bodySmall, color: IvoryMuted)
                        }
                        .frame(maxWidth: .infinity)
                        .padding(.top, 16)

                        HairlineButton(text: "Invite friends · \(group.inviteCode)", icon: RollIcon.personAdd) {
                            Sharing.shareInvite(groupName: group.name, code: group.inviteCode, link: group.inviteLink)
                        }
                        .padding(.horizontal, 20).padding(.vertical, 12)
                    }

                    ForEach(state.members) { member in
                        MemberRow(
                            member: member,
                            isMe: member.uid == state.myUid,
                            canManage: state.isAdmin && member.uid != state.myUid,
                            onRemove: { confirmRemoval(member) }
                        )
                    }
                }
                .padding(.bottom, 32)
            }
        }
        .background(Ink.ignoresSafeArea())
    }

    private func confirmRemoval(_ member: Member) {
        overlays.confirm(
            title: "Remove \(member.name)?",
            message: "They'll lose access to this roll's photos. Photos they already uploaded stay in the roll.",
            confirmLabel: "Remove"
        ) {
            viewModel.removeMember(member.uid)
        }
    }
}

private struct MemberRow: View {
    let member: Member
    let isMe: Bool
    let canManage: Bool
    let onRemove: () -> Void

    var body: some View {
        HStack(spacing: 0) {
            UserAvatar(name: member.name, photoUrl: member.photoUrl, size: 44, seed: member.uid)
            Spacer().frame(width: 14)
            VStack(alignment: .leading, spacing: 2) {
                HStack(spacing: 8) {
                    Text(isMe ? "\(member.name) (you)" : member.name).rollStyle(RollType.titleMedium, color: Ivory)
                    if member.isAdmin {
                        Text("Admin")
                            .rollStyle(RollType.labelSmall, color: Ivory)
                            .padding(.horizontal, 7).padding(.vertical, 2)
                            .background(RoundedRectangle(cornerRadius: RollShapes.extraSmall).fill(GoldDeep))
                    }
                }
                Text("\(member.photoCount) \(member.photoCount == 1 ? "photo" : "photos")")
                    .rollStyle(RollType.bodyMedium, color: IvoryMuted)
            }
            .frame(maxWidth: .infinity, alignment: .leading)

            if canManage {
                Menu {
                    Button("Remove from roll", role: .destructive, action: onRemove)
                } label: {
                    Image(systemName: RollIcon.more)
                        .font(.system(size: 19, weight: .semibold))
                        .foregroundStyle(Ivory)
                        .frame(width: 48, height: 48)
                        .contentShape(Circle())
                }
                .accessibilityLabel("Manage \(member.name)")
            }
        }
        .padding(.horizontal, 16).padding(.vertical, 10)
    }
}
